File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []

        letterMap = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        def backtrack(start, cur):
            if len(cur) == len(digits):
                ans.append("".join(cur))
                return

            for i in range(start, len(digits)):
                digit = digits[i]
                for letter in letterMap[digit]:
                    cur.append(letter)
                    backtrack(i + 1, cur)
                    cur.pop()

        ans = []
        backtrack(0, [])
        return ans

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.php
class Solution {

    /**
     * @param String $digits
     * @return String[]
     */
    function letterCombinations($digits) {
        $letterMapping = [
            '2' => ['a','b','c'],
            '3' => ['d','e','f'],
            '4' => ['g','h','i'],
            '5' => ['j','k','l'],
            '6' => ['m','n','o'],
            '7' => ['p','q','r','s'],
            '8' => ['t','u','v'],
            '9' => ['w','x','y','z'],
        ];
        if(empty($digits)){
            return [];
        }

        $ans = [];
        $backtrack = function($comb,$index)use(&$ans,&$backtrack,$letterMapping,$digits){
            if($index === strlen($digits)){
                $ans[] = implode("",$comb);
                return;
            }

            $digit = $digits[$index];
            foreach($letterMapping[$digit] as $letter){
                $comb[] = $letter;
                $backtrack($comb,$index+1);
                array_pop($comb);
            }
        };

        $backtrack([],0);
        return $ans;
    }
}